app/ratelimit.go

```diff
@@ -0,0 +1,131 @@
+// Copyright (c) 2018-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+package app
+
+import (
+	"math"
+	"net/http"
+	"strconv"
+	"strings"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/mattermost-server/model"
+	"github.com/mattermost/mattermost-server/utils"
+	throttled "gopkg.in/throttled/throttled.v2"
+	"gopkg.in/throttled/throttled.v2/store/memstore"
+)
+
+type RateLimiter struct {
+	throttledRateLimiter *throttled.GCRARateLimiter
+	useAuth              bool
+	useIP                bool
+	header               string
+}
+
+func NewRateLimiter(settings *model.RateLimitSettings) *RateLimiter {
+	store, err := memstore.New(*settings.MemoryStoreSize)
```

Comment:
Is this memory store external to Mattermost? Just wondering how this works in clustered environments where there could be more than one Mattermost instance. If it's an in-memory cache, the rate limiting may not work as expected

Reply:
In a clustered environment rate limiting would be per-server. There is the ability to use redis as a store. So when we move to using that, we might move this to redis for large customers. But I am not sure the perf trade off for the sake of accuracy would be worth it.